Geometry_Engine/Compute/OutlinesFromLines.cs

```diff
@@ -0,0 +1,427 @@
+            {
+                // Start from arbitrary node and edge
+                Point startNode = graph.Keys.First();
+                Line currentEdge = graph[startNode].First();
+                graph[startNode].Remove(currentEdge);
+                if (graph[startNode].Count == 0)
+                    graph.Remove(startNode);
+
+                Vector currentDir = dirs[currentEdge];
+                Point currentNode = currentEdge.End;
+
+                // Keep on turning left until hitting the start node
+                List<Line> outline = new List<Line> { currentEdge };
+                while (currentNode != startNode)
+                {
+                    List<Vector> nodeDirs = graph[currentNode].Select(x => dirs[x]).ToList();
+                    Vector newDir = null;
+                    Line newEdge = null;
+
+                    // To avoid computing heavy signed angle, first check only dirs to the left from the current dir
+                    Vector headingLeft = new Vector { X = -currentDir.Y, Y = currentDir.X };
+                    for (int j = 0; j < nodeDirs.Count; j++)
+                    {
+                        // Avoid the edge itself as well as the same one in opposite direction
+                        if (graph[currentNode][j] == currentEdge || 1 + currentDir.DotProduct(nodeDirs[j]) < sqTol)
+                            continue;
+
+                        if (headingLeft.DotProduct(nodeDirs[j]) < 0)
+                            continue;
+
+                        if (newDir == null || nodeDirs[j].DotProduct(currentDir) < newDir.DotProduct(currentDir))
+                        {
+                            newDir = nodeDirs[j];
+                            newEdge = graph[currentNode][j];
+                        }
+                    }
+
+                    // If not found (and only if), then look up the ones to the right
+                    if (newEdge == null)
+                    {
+                        for (int j = 0; j < nodeDirs.Count; j++)
+                        {
+                        // Avoid the edge itself as well as the same one in opposite direction
+                            if (graph[currentNode][j] == currentEdge || 1 + currentDir.DotProduct(nodeDirs[j]) < sqTol)
+                                continue;
+
+                            if (headingLeft.DotProduct(nodeDirs[j]) >= 0)
+                                continue;
+
+                            if (newDir == null || nodeDirs[j].DotProduct(currentDir) > newDir.DotProduct(currentDir))
+                            {
+                                newDir = nodeDirs[j];
+                                newEdge = graph[currentNode][j];
+                            }
+                        }
+                    }
+
+                    if (newEdge == null)
+                    {
+                        BH.Engine.Base.Compute.RecordError("Outlines could not be created due to an internal error.");
```

Comment:
It is a fallback, I would say this error should never be hit, but knowing the reality I put a generic message with return `null` to avoid a more serious crash.